File: Kern/Sources/DuoKern/Box/Growth.swift
import Foundation

// Growth rules: day keys, direction-scoped scheduling access, health gate,
// new-card budget, phrase unlock, and deterministic candidate ordering.

extension BoxEngine {

    // MARK: - Day keys

    /// Day key = startOfDay in the caller's calendar, formatted yyyy-MM-dd.
    static func dayKey(for now: Date, calendar: Calendar) -> String {
        let start = calendar.startOfDay(for: now)
        let c = calendar.dateComponents([.year, .month, .day], from: start)
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    // MARK: - Direction-scoped scheduling access

    /// Scheduling for a card in the current direction, if any.
    static func scheduling(_ state: BoxState, _ cardID: String) -> CardScheduling? {
        state.scheduling[BoxState.schedulingKey(cardID: cardID, direction: state.config.direction)]
    }

    /// All scheduling entries for the current direction whose card still exists.
    static func directionSchedulings(_ state: BoxState) -> [CardScheduling] {
        state.scheduling.values.filter {
            $0.direction == state.config.direction && state.cards[$0.cardID] != nil
        }
    }

    /// Current-direction, non-suspended scheduling entries.
    static func activeSchedulings(_ state: BoxState) -> [CardScheduling] {
        directionSchedulings(state).filter { !$0.suspended }
    }

    /// Active entries with `due <= now`, oldest due first, ties broken by card id.
    static func dueSchedulings(_ state: BoxState, now: Date) -> [CardScheduling] {
        activeSchedulings(state)
            .filter { ($0.due ?? .distantFuture) <= now }
            .sorted { ($0.due ?? now, $0.cardID) < ($1.due ?? now, $1.cardID) }
    }

    // MARK: - Health gate & budget

    /// Design §Box 2: introduce new cards only if
    /// (a) projected post-session backlog < dueSoftCap, and
    /// (b) relearning share < 20% — sub-gate applies only once activeCount ≥ 10
    ///     (else it passes; day-one bootstrap).
    static func healthGateOpen(_ state: BoxState, now: Date) -> Bool {
        let active = activeSchedulings(state)
        let dueCount = active.filter { ($0.due ?? .distantFuture) <= now }.count
        let projectedBacklog = dueCount - min(dueCount, state.config.sessionCap)
        guard projectedBacklog < state.config.dueSoftCap else { return false }
        if active.count >= 10 {
            let relearning = active.filter { $0.phase == .relearning }.count
            guard Double(relearning) < 0.2 * Double(active.count) else { return false }
        }
        return true
    }

    /// Load-based budget: free slots in the learning pool, ignoring the health
    /// gate. Growth tops the `.learning` pool back up to `maxLearning` — so a
    /// day where cards graduate frees room for more, and a full pool adds none.
    static func learningPoolBudget(_ state: BoxState) -> Int {
        let learning = activeSchedulings(state).filter { $0.phase == .learning }.count
        return max(0, state.config.maxLearning - learning)
    }

    /// Budget after the health gate: 0 when the gate is closed.
    static func gatedNewBudget(_ state: BoxState, now: Date) -> Int {
        healthGateOpen(state, now: now) ? learningPoolBudget(state) : 0
    }

    // MARK: - Phrase unlock (design §Box 3)

    /// A component counts as stable only with current-direction scheduling in
    /// `.review` phase, not suspended, and stability ≥ phraseUnlockStability.
    /// Missing scheduling or suspended = not stable.
    static func isComponentStable(_ state: BoxState, _ componentID: String) -> Bool {
        guard let sched = scheduling(state, componentID),
              !sched.suspended,
              sched.phase == .review,
              let memory = sched.memory else { return false }
        return memory.stability >= state.config.phraseUnlockStability
    }

    /// Unlock fast path: phrases with ZERO components never take it
    /// (they follow normal seed order instead).
    static func isPhraseUnlocked(_ state: BoxState, _ card: Card) -> Bool {
        guard card.kind == .phrase, !card.componentIDs.isEmpty else { return false }
        return card.componentIDs.allSatisfy { isComponentStable(state, $0) }
    }

    // MARK: - Deterministic seed order

    /// Nouns/verbs sort ahead of phrases within an area (design §Box 5).
    static func kindRank(_ kind: CardKind) -> Int {
        switch kind {
        case .noun: return 0
        case .verb: return 1
        case .phrase: return 2
        }
    }

    /// Curated import order via `seedIndex`; ties (hand-built cards default to 0)
    /// fall back to area name, nouns/verbs before phrases, then card id.
    static func seedOrdered(_ cards: some Sequence<Card>) -> [Card] {
        cards.sorted {
            ($0.seedIndex, $0.area, kindRank($0.kind), $0.id)
                < ($1.seedIndex, $1.area, kindRank($1.kind), $1.id)
        }
    }

    // MARK: - New-card candidates

    /// Enqueued ids that could actually enter now: unscheduled, and — for
    /// phrases with components — unlocked. User agency: these BYPASS the budget.
    static func enqueuedEligible(_ state: BoxState) -> [String] {
        state.enqueued.filter { id in
            guard let card = state.cards[id], scheduling(state, id) == nil else { return false }
            return card.kind != .phrase || card.componentIDs.isEmpty || isPhraseUnlocked(state, card)
        }
    }

    /// Candidate selection, bounded by `capacity` total introductions:
    /// 1. Enqueued eligible ids (queue order) — the user asked for them, so they
    ///    bypass `budget` (matching `answer()` and the extra round).
    /// 2. Automatic growth within `budget`: unlocked phrases (fast path), then
    ///    seed-order words. Locked phrases never enter, even enqueued.
    static func newCandidates(_ state: BoxState, budget: Int, capacity: Int)
        -> (unlockedPhrases: [String], newWords: [String]) {
        guard capacity > 0 else { return ([], []) }

        let unscheduled = state.cards.values.filter { scheduling(state, $0.id) == nil }
        var taken = Set<String>()
        var slots = capacity

        // 1. Enqueued — bypass the budget, bounded only by session capacity.
        var words: [String] = []
        for id in enqueuedEligible(state) where slots > 0 {
            guard !taken.contains(id) else { continue }
            words.append(id)
            taken.insert(id)
            slots -= 1
        }

        // 2. Automatic growth within the (load-based) budget.
        var autoRemaining = min(budget, slots)
        let phrases = Array(
            seedOrdered(unscheduled.filter { !taken.contains($0.id) && isPhraseUnlocked(state, $0) })
                .map(\.id).prefix(autoRemaining))
        phrases.forEach { taken.insert($0) }
        autoRemaining -= phrases.count

        for card in seedOrdered(unscheduled) {
            guard autoRemaining > 0 else { break }
            guard !taken.contains(card.id) else { continue }
            if card.kind == .phrase, !card.componentIDs.isEmpty, !isPhraseUnlocked(state, card) {
                continue // locked phrase: waits for its components
            }
            words.append(card.id)
            taken.insert(card.id)
            autoRemaining -= 1
        }
        return (phrases, words)
    }
}

File: kmp/kern/src/commonMain/kotlin/net/spross/kern/box/Growth.kt
package net.spross.kern.box

import kotlin.math.max
import kotlin.math.min
import kotlin.time.Instant
import net.spross.kern.model.Card
import net.spross.kern.model.CardKind
import net.spross.kern.model.CardPhase
import net.spross.kern.model.ExerciseUnit
import net.spross.kern.model.Role
import net.spross.kern.model.UnitKey

/** New-unit candidates: automatic unlock fast-path phrases separate from the rest. */
internal data class NewCandidates(
    val unlockedPhrases: List<String>,
    val newUnits: List<String>,
) {
    companion object {
        val empty = NewCandidates(emptyList(), emptyList())
    }
}

internal object Growth {

    /**
     * Health gate (denominated in UNITS): projected post-session backlog < dueSoftCap,
     * and relearning share < 20 % of active units — the sub-gate applies only once
     * active >= 10 (else it passes; day-one bootstrap).
     */
    fun healthGateOpen(state: BoxState, nowEpochMillis: Long): Boolean {
        val now = Instant.fromEpochMilliseconds(nowEpochMillis)
        val active = Inventory.active(state)
        val dueCount = active.count { it.due != null && it.due <= now }
        val projectedBacklog = dueCount - min(dueCount, state.config.sessionCap)
        if (projectedBacklog >= state.config.dueSoftCap) return false
        if (active.size >= 10) {
            val relearning = active.count { it.phase == CardPhase.Relearning }
            if (relearning >= 0.2 * active.size) return false
        }
        return true
    }

    /** Free learning-pool slots in CONCEPTS, ignoring the health gate. */
    fun learningPoolBudget(state: BoxState): Int =
        max(0, state.config.maxLearning - Inventory.conceptsInFlight(state).size)

    /** Budget after the health gate: 0 when the gate is closed. */
    fun gatedNewBudget(state: BoxState, nowEpochMillis: Long): Int =
        if (healthGateOpen(state, nowEpochMillis)) learningPoolBudget(state) else 0

    // Phrase unlock reads `id|produce` schedules RAW BY KEY — join- and source-independent,
    // so a source switch can never re-lock a phrase.
    fun isComponentStable(state: BoxState, componentId: String): Boolean {
        val sched = state.scheduling[UnitKey.produce(componentId).encoded] ?: return false
        if (sched.suspended || sched.phase != CardPhase.Review) return false
        return (sched.memory?.stability ?: 0.0) >= state.config.phraseUnlockStability
    }

    /** Zero-component phrases never take the fast path (they follow seed order). */
    fun isPhraseUnlocked(state: BoxState, card: Card): Boolean =
        card.kind == CardKind.Phrase &&
            card.components.isNotEmpty() &&
            card.components.all { isComponentStable(state, it) }

    /** Whether the card may enter at all: word, component-free phrase, or unlocked phrase. */
    private fun cardIntroducible(state: BoxState, card: Card): Boolean =
        card.kind != CardKind.Phrase || card.components.isEmpty() || isPhraseUnlocked(state, card)

    /**
     * Whether an unscheduled unit may auto-enter. Produce follows card eligibility;
     * recognize waits for its card's produce unit to sit in Review un-suspended
     * (eligibility lag — kills first-rating echo and halves day-one pool pressure).
     */
    fun isIntroducible(state: BoxState, unit: ExerciseUnit): Boolean = when (unit.role) {
        Role.Produce -> cardIntroducible(state, unit.card)
        Role.Recognize -> {
            val produce = state.scheduling[UnitKey.produce(unit.card.id).encoded]
            produce != null && !produce.suspended && produce.phase == CardPhase.Review
        }
    }

    /** Enqueued concept ids that could enter now: joined, produce-unscheduled, not locked. */
    fun enqueuedEligible(state: BoxState): List<String> = state.enqueued.filter { id ->
        val card = state.cards[id] ?: return@filter false
        state.scheduling[UnitKey.produce(id).encoded] == null && cardIntroducible(state, card)
    }

    /**
     * Candidate selection in unit keys. `conceptBudget` caps how many concepts may ENTER
     * the learning pool; units of concepts already in flight ride free (they do not grow
     * the pool). Enqueued produce units lead and bypass the health gate; with the gate
     * open, unlocked phrases enter next, then seed-order units (recognize backfill of
     * graduated material naturally leads new produce — lower seed index). `capacityUnits`
     * caps the total in units.
     */
    fun newCandidates(
        state: BoxState,
        conceptBudget: Int,
        gateOpen: Boolean,
        capacityUnits: Int,
    ): NewCandidates {
        var capacity = capacityUnits
        var conceptSlots = conceptBudget
        if (capacity <= 0) return NewCandidates.empty

        val inFlight = Inventory.conceptsInFlight(state).toMutableSet()
        val takenKeys = mutableSetOf<String>()
        val unlockedPhrases = mutableListOf<String>()
        val newUnits = mutableListOf<String>()

        // 1. Enqueued lead — within the concept throttle, bypassing the health gate.
        for (id in enqueuedEligible(state)) {
            if (capacity <= 0 || conceptSlots <= 0) break
            val key = UnitKey.produce(id).encoded
            if (key in takenKeys) continue
            newUnits += key
            takenKeys += key
            inFlight += id
            conceptSlots -= 1
            capacity -= 1
        }
        if (!gateOpen) return NewCandidates(unlockedPhrases, newUnits)

        val unscheduled = Inventory.joinedUnits(state).filter { state.scheduling[it.key] == null }

        // 2a. Unlock fast path: component phrases whose components all sit stable.
        for (unit in unscheduled) {
            if (capacity <= 0 || conceptSlots <= 0) break
            if (unit.role != Role.Produce || unit.card.kind != CardKind.Phrase) continue
            if (unit.card.components.isEmpty() || unit.key in takenKeys) continue
            if (!isPhraseUnlocked(state, unit.card)) continue
            unlockedPhrases += unit.key
            takenKeys += unit.key
            inFlight += unit.card.id
            conceptSlots -= 1
            capacity -= 1
        }

        // 2b. Automatic seed-order growth (locked phrases wait for their components).
        for (unit in unscheduled) {
            if (capacity <= 0) break
            if (unit.key in takenKeys || !isIntroducible(state, unit)) continue
            val ridesFree = unit.card.id in inFlight
            if (!ridesFree && conceptSlots <= 0) continue
            newUnits += unit.key
            takenKeys += unit.key
            if (!ridesFree) {
                inFlight += unit.card.id
                conceptSlots -= 1
            }
            capacity -= 1
        }
        return NewCandidates(unlockedPhrases, newUnits)
    }
}
